crates/developer/src/main.rs

```diff
@@ -0,0 +1,1086 @@
+use tracing_appender::rolling::{RollingFileAppender, Rotation};
+use tracing_subscriber::{self, EnvFilter};
+
+pub struct DeveloperRouter {
+    tools: Vec<Tool>,
+    // The cwd, active_resources, and file_history are shared across threads
+    // so we need to use an Arc to ensure thread safety
+    cwd: Arc<Mutex<PathBuf>>,
+    active_resources: Arc<Mutex<HashMap<String, Resource>>>,
+    file_history: Arc<Mutex<HashMap<PathBuf, Vec<String>>>>,
+    instructions: String,
+}
+
+impl DeveloperRouter {
+    pub fn new() -> Self {
+        let bash_tool = Tool::new(
+            "bash".to_string(),
+            "Run a bash command in the shell in the current working directory".to_string(),
+            json!({
+                "type": "object",
+                "required": ["command"],
+                "properties": {
+                    "command": {"type": "string"}
+                }
+            }),
+        );
+
+        let text_editor_tool = Tool::new(
+            "text_editor".to_string(),
+            "Perform text editing operations on files.".to_string(),
+            json!({
+                "type": "object",
+                "required": ["command", "path"],
+                "properties": {
+                    "path": {"type": "string"},
+                    "command": {
+                        "type": "string",
+                        "enum": ["view", "write", "str_replace", "undo_edit"]
+                    },
+                    "new_str": {"type": "string"},
+                    "old_str": {"type": "string"},
+                    "file_text": {"type": "string"}
+                }
+            }),
+        );
+
+        let instructions = "Developer instructions...".to_string(); // Reuse from original code
+
+        let cwd = std::env::current_dir().unwrap();
+        let mut resources = HashMap::new();
+        let uri = format!("str:///{}", cwd.display());
+        let resource = Resource::new(
+            uri.clone(),
+            Some("text".to_string()),
+            Some("cwd".to_string()),
+        )
+        .unwrap();
+        resources.insert(uri, resource);
+
+        Self {
+            tools: vec![bash_tool, text_editor_tool],
+            cwd: Arc::new(Mutex::new(cwd)),
+            active_resources: Arc::new(Mutex::new(resources)),
+            file_history: Arc::new(Mutex::new(HashMap::new())),
+            instructions,
+        }
+    }
+
+    // Example utility function to call the underlying logic
+    async fn call_bash(&self, args: Value) -> Result<Value, ToolError> {
+        let result = self.bash(args).await; // adapt your logic from DeveloperSystem
+        self.map_agent_result_to_value(result)
+    }
+
+    async fn call_text_editor(&self, args: Value) -> Result<Value, ToolError> {
+        let result = self.text_editor(args).await; // adapt from DeveloperSystem
+        self.map_agent_result_to_value(result)
+    }
+
+    // Convert AgentResult<Vec<Content>> to Result<Value, ToolError>
```

Comment:
nit: i haven't thought about it too hard, but might be better to just rework the old code a bit to create ToolError?